### PR TITLE
Split tests into groups + SafeTestsets

### PR DESCRIPTION
**Pull request by lkdvos**
This PR reorganizes the test structure as described in #28 .

The grouping logic happens automatically based on file structure:
- Tests without a group, in the top folder are run as part of the group "top" 
`test/mytest1.jl`
`test/mytest2.jl`
- Tests within a folder are assigned to the group based on the folder name:
`test/utility/test1.jl`
`test/basics/test2.jl`
- Examples are run as part of the group "examples"
`examples/myexample.jl`

To select/run a subset of the tests, the `Tests.yml` CI can make use of the `group` input. Selecting this in a `matrix` ensures parallelism through the github runners.

To select/run a subset of the tests locally, the `test_args` keyword, along with the `--group=groupname` syntax can be used, for example:
```julia
using Pkg: Pkg
Pkg.test("mypackage"; test_args=`--group=examples`)
```

Note that all group name matches are insensitive to capitalization. The default group `ALL` runs all tests.

### CODE REVIEW

**Comment by mtfishman** [comment]
Nice, I really like the idea of basing the groups off of the folder structure!

**Comment by mtfishman** [comment]
Does this also work with the environment variable? Or did you decide not to go that route?

File: templates/test/test/runtests.jl [hunk 1]
```diff
@@ -1,12 +1,36 @@
-@eval module $(gensym())
-using Test: @testset
+using SafeTestsets
```
Comment (mtfishman) [inline templates/test/test/runtests.jl:1]:
```suggestion
using SafeTestsets: @safetestset
```

File: templates/test/test/runtests.jl [hunk 2]
```diff
@@ -1,12 +1,36 @@
...
+arg_id = findfirst(contains(pat), ARGS)
+const GROUP = uppercase(
+  if isnothing(arg_id)
+    get(ENV, "GROUP", "ALL")
+  else
+    only(match(pat, ARGS[arg_id]).captures)
+  end,
+)
+
+@time begin
```
Comment (mtfishman) [inline templates/test/test/runtests.jl:16]: Doesn't `@testset` report times now?

Comment (mtfishman) [reply templates/test/test/runtests.jl:16]: I.e. could we achieve this by wrapping all of this in `@testset` or `@safetestset`?

Comment (lkdvos) [reply templates/test/test/runtests.jl:16]: I'm not entirely sure, I mostly copied this from my other workflows. I think the main difference is that `@time` also reports splits in compiletime/runtime, which I find interesting, but I have absolutely no preference about this.
Just as an aside, wrapping things in nested `@testset` structures makes it such that output is only printed after all tests finish. Not a big deal, but something to keep in mind.

**Comment by lkdvos** [comment]
This works both with environment variables and test args, it first checks the latter, if this is empty it checks the environment variable, if that is empty it runs all tests

**Comment by lkdvos** [comment]
(TODO: codecov token)

**Comment by lkdvos** [comment]
If the tests pass, I would merge this and I'll add the setup folder/files in a separate PR.

**Pull request merged**